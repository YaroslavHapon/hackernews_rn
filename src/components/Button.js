import React from 'react'
import { View, Text, TouchableOpacity  } from 'react-native'
import PropTypes from 'prop-types'

const Button = ({ text, onPress, style, buttonTextStyle}) => {
  return(
    <TouchableOpacity onPress={onPress}>
      <View style={style}>
        <Text style={buttonTextStyle}>{text}</Text>
      </View>
    </TouchableOpacity>
  )
};
Button.propTypes = {
  text: PropTypes.string,
  onPress: PropTypes.func,
  style: PropTypes.number,
  buttonTextStyle: PropTypes.number
}
export default Button