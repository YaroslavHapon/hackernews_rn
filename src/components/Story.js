import React from 'react'
import { View, Text, Linking, ActivityIndicator, StyleSheet, TouchableHighlight } from 'react-native'
import Button from './Button'
import PropTypes from 'prop-types'

const styles = StyleSheet.create({
  storyContainer: {
    flex: 1,
    padding: 10,
    backgroundColor: "#fff"
  },
  loadingContainer: {
    flex: 1,
    paddingTop: 20,
    height: 120
  },
  separator: {
    height: 1,
    backgroundColor: "#CED0CE",
    marginLeft: 10,
    marginRight: 10
  },
  button: {
    alignItems: 'center',
    backgroundColor: "#2233aa",
    marginTop: 15,
    marginBottom: 15
  },
  title: {
    color: '#333',
    fontWeight: 'bold',
    fontSize: 18
  },
  storyInfo: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center'
  },
  storyInfoText: {
    fontSize: 12
  },
  buttonTextStyle: {
    color: '#fff',
    padding: 5
  }
});

export default class Story extends React.Component{
  static propTypes = {
    fetchStory: PropTypes.func,
    storyGet: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.shape({
        loading: PropTypes.bool,
        loaded: PropTypes.bool,
        data: PropTypes.object,
        error: PropTypes.object
      })
    ]),
    navigation: PropTypes.object
  }

  componentWillMount () {
    this.props.fetchStory(this.props.story.item);
  }

  openUrl = url => {
    Linking.openURL(url)
  }

  onShowComments = () => {
    const { story, storyGet } = this.props;
    this.props.navigate('CommentList', { data: storyGet[story.item].data });
  }

  render () {
    const { storyGet, story } = this.props

    if(!storyGet[story.item]) {
      return <View style={styles.loadingContainer} />;
    }

    if (storyGet[story.item].loading) {
      return (
        <View style={styles.loadingContainer}>
          <ActivityIndicator />
        </View>
      );
    }

    if (storyGet[story.item].error) {
      return (
        <View>{storyGet[story.item].error.message}</View>
      );
    }
    
    return(
      <View>
        {storyGet[story.item].data &&
          <TouchableHighlight onPress={this.onShowComments} underlayColor="#ccc">
            <View style={styles.storyContainer}>
              <Text style={styles.title}>{storyGet[story.item].data.title}</Text>
              <View style={styles.storyInfo}>
                <Text style={styles.storyInfoText}>
                  <Text>Author: </Text>
                  {storyGet[story.item].data.by}
                </Text>
                <Text style={styles.storyInfoText}>
                  <Text>Score: </Text>
                  {storyGet[story.item].data.score}
                </Text>
              </View>
              <Button
                style={styles.button}
                buttonTextStyle={styles.buttonTextStyle}
                text="Open In Browser"
                onPress={() => this.openUrl(storyGet[story.item].data.url)} />
            </View>
          </TouchableHighlight>
        }
        <View style={styles.separator}/>
      </View>
    )
  }
}