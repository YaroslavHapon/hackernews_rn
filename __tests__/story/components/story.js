import { ActivityIndicator, View, TouchableHighlight,
  Text, Linking  } from  'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon'
import Button from './../../../src/components/Button'
import Story from './../../../src/components/Story'

jest.mock('Linking', () => {
  return {
    openURL: jest.fn()
  }
});

const mockProps = {
  story: { item: 1234 },
  storyGet: {},
  fetchStory: sinon.spy()
};

describe('Testing Story component', () => {
  const wrapper = shallow(
    <Story {...mockProps} />
  );

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
    wrapper.setProps({ story: { item: 5678 }});
    expect(wrapper).toMatchSnapshot();
  });

  it('should call the fetchStory function', () => {
    expect(wrapper.instance().props.fetchStory.calledOnce).toBe(true);
    expect(wrapper.instance().props.fetchStory.args[0][0]).toBe(1234);
  });

  it('should show ActivityIndicator, while request loading', () => {
    const wrapper = shallow(
      <Story {...mockProps} storyGet= {{ 1234: { loading: true }} } />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(ActivityIndicator).length).toBe(1);
  });

  it('should show ActivityIndicator, when request fails', () => {
    const error = { message: 'Request error!' };
    const wrapper = shallow(
      <Story {...mockProps} storyGet= {{ 1234: { error: error }}} />
    );
    expect(wrapper.find(View).props().children).toBe(error.message);
  });

  const data = {
    title: 'New story',
    by: 'Jhon Doe',
    score: 100,
    url: 'http://url'
  };

  it('should render with data', () => {
    const wrapperWithData = shallow(
      <Story {...mockProps}
        storyGet={ { 1234: { data: data }}}
      />
    );
    expect(wrapperWithData).toMatchSnapshot();
  });

  it('should navigate to comments, on touch', () => {
    const navigate = sinon.spy();
    const wrapperWithData = shallow(
      <Story {...mockProps}
        storyGet={ { 1234: { data: data }}}
        navigate={navigate}
      />
    );

    wrapperWithData.find(TouchableHighlight).props().onPress();
    expect(navigate.calledOnce).toBe(true);
  });

  it('should show the right data', () => {
    const wrapperWithData = shallow(
      <Story {...mockProps}
        storyGet={ { 1234: { data: data }}}
      />
    );
    expect(wrapperWithData).toMatchSnapshot();
    expect(wrapperWithData.find(Text).nodes[0].props.children)
      .toBe(data.title);
    expect(wrapperWithData.find(Text).nodes[1].props.children[1])
      .toBe(data.by);
    expect(wrapperWithData.find(Text).nodes[3].props.children[1])
      .toBe(data.score)
  });

  it('should open the story in browser', () => {
    const wrapperWithData = shallow(
      <Story {...mockProps}
        storyGet={ { 1234: { data: data }}}
      />
    );
    const openURlSpy = sinon.spy(Linking, 'openURL');
    wrapperWithData.find(Button).props().onPress();
    expect(openURlSpy.calledOnce).toBe(true);
  });
});
