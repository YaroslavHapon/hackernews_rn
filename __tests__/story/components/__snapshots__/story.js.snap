// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Story component should render correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "height": 120,
      "paddingTop": 20,
    }
  }
/>
`;

exports[`Testing Story component should render correctly 2`] = `
<View
  style={
    Object {
      "flex": 1,
      "height": 120,
      "paddingTop": 20,
    }
  }
/>
`;

exports[`Testing Story component should render with data 1`] = `
<View>
  <TouchableHighlight
    activeOpacity={0.85}
    onPress={[Function]}
    underlayColor="#ccc"
  >
    <View
      style={
        Object {
          "backgroundColor": "#fff",
          "flex": 1,
          "padding": 10,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#333",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        New story
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 12,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            Author: 
          </Text>
          Jhon Doe
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 12,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            Score: 
          </Text>
          100
        </Text>
      </View>
      <Button
        buttonTextStyle={
          Object {
            "color": "#fff",
            "padding": 5,
          }
        }
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2233aa",
            "marginBottom": 15,
            "marginTop": 15,
          }
        }
        text="Open In Browser"
      />
    </View>
  </TouchableHighlight>
  <View
    style={
      Object {
        "backgroundColor": "#CED0CE",
        "height": 1,
        "marginLeft": 10,
        "marginRight": 10,
      }
    }
  />
</View>
`;

exports[`Testing Story component should show ActivityIndicator, while request loading 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "height": 120,
      "paddingTop": 20,
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;

exports[`Testing Story component should show the right data 1`] = `
<View>
  <TouchableHighlight
    activeOpacity={0.85}
    onPress={[Function]}
    underlayColor="#ccc"
  >
    <View
      style={
        Object {
          "backgroundColor": "#fff",
          "flex": 1,
          "padding": 10,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#333",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        New story
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 12,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            Author: 
          </Text>
          Jhon Doe
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 12,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            Score: 
          </Text>
          100
        </Text>
      </View>
      <Button
        buttonTextStyle={
          Object {
            "color": "#fff",
            "padding": 5,
          }
        }
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2233aa",
            "marginBottom": 15,
            "marginTop": 15,
          }
        }
        text="Open In Browser"
      />
    </View>
  </TouchableHighlight>
  <View
    style={
      Object {
        "backgroundColor": "#CED0CE",
        "height": 1,
        "marginLeft": 10,
        "marginRight": 10,
      }
    }
  />
</View>
`;
